### PR TITLE
Current version of the project.

### PR DESCRIPTION
**Pull request by kmavropulo**

### CODE REVIEW

File: src/main/java/com/github/admissionCommittee/util/DummyDatabaseInit.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                "user@epam.com", "123", LocalDate.of(1990, Month.APRIL, 2));
+        final User admin1 = new User(UserTypeEnum.ADMIN, "Семёнова", "Юлия", "Викторовна",
+                "iuliia_semionova@epam.com", "123456", LocalDate.of(1985, Month.DECEMBER, 22));
+        final User user2 = new User(UserTypeEnum.USER, "Касьянов", "Максим", "Евгеньевич",
+                "kasianov_maksim@epam.com", "PaSsWoRd", LocalDate.of(1997, Month.FEBRUARY, 17));
+
+        userDao.create(user1);
+        userDao.create(admin1);
+        userDao.create(user2);
+        userDao.create(user3);
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/util/DummyDatabaseInit.java]: you can inline it, still bad naming

File: src/main/java/com/github/admissionCommittee/util/DummyDatabaseInit.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+
+        SubjectDao subjectDao = DaoFactory.getDaoFactory().getSubjectDao();
+
+        final Subject subject1 = new Subject(SubjectNameEnum.BIOLOGY);
+        final Subject subject2 = new Subject(SubjectNameEnum.ALGEBRA);
+        final Subject subject3 = new Subject(SubjectNameEnum.RUSSIAN);
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/util/DummyDatabaseInit.java]: this variables totally can be named better

File: build.gradle [hunk 3]
```diff
@@ -1,16 +1,34 @@
...
+
+    compile 'ch.qos.logback:logback-classic:+',
+            'com.h2database:h2:+',
+            'org.glassfish.web:javax.servlet.jsp.jstl:+',
+            'org.glassfish:javax.el:3.0.1-b08'
+
+    compile group: 'org.hibernate', name: 'hibernate-core', version: '5.2.12.Final'
+    compile group: 'org.hibernate.validator', name: 'hibernate-validator', version: '6.0.4.Final'
+    compile group: 'org.projectlombok', name: 'lombok', version: '1.16.18'
+
```
Comment (SMishAlex) [inline build.gradle]: see previous review

File: src/main/java/com/github/admissionCommittee/dao/DaoFactory.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    private DaoFactory() {
+    }
+
+    private static ExamCertificateDao examCertificateDao;
+    private static FacultyDao facultyDao;
+    private static SchoolCertificateDao schoolCertificateDao;
+    private static SheetDao sheetDao;
+    private static SubjectDao subjectDao;
+    private static UserDao userDao;
+    private static volatile DaoFactory daoFactory;
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/dao/DaoFactory.java:13]: it usually contains word instance, the same for getter

File: src/main/java/com/github/admissionCommittee/dao/GenericDao.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    @SuppressWarnings("unchecked")
+    public List<T> getAll() {
+        openSessionWithTransaction();
+        List<T> instances = (List<T>) session.createQuery("from " + type
+                .getName()).list();
+        closeSessionWithTransaction();
+        return instances;
+    }
+
+    public void delete(long id) {
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/dao/GenericDao.java:64]: double db call and double transaction in one dao method....

File: src/main/java/com/github/admissionCommittee/web/controllers/AdminController.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+@WebServlet({"/admin"})
+public class AdminController extends HttpServlet {
+    private static final Logger log = LoggerFactory.getLogger(LoginController.class);
+    private UserService service;
+
+    @Override
+    public void init(ServletConfig config) throws ServletException {
+        service = (UserService) config.getServletContext().getAttribute("userService");
+    }
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/web/controllers/AdminController.java]: it's better to call also super.init() when you override something in standard library

File: src/main/java/com/github/admissionCommittee/web/controllers/filters/AuthorizationFilter.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+        HttpSession session = httpRequest.getSession(true);
+        if (session.getAttribute("user_id") == null) {
+            log.info(String.format("Session %s not authorized, redirect to /", session.getId()));
+            httpResponse.sendRedirect("/");
+            return;
+        }
+
+        chain.doFilter(request,response);
+        log.trace("Quit from User authorization filter");
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/web/controllers/filters/AuthorizationFilter.java]: you will see this log after controller (it calls in doFilter after last filter)

File: src/main/java/com/github/admissionCommittee/web/controllers/listeners/InitListener.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.ServletContextEvent;
+import javax.servlet.ServletContextListener;
+import javax.servlet.annotation.WebListener;
+
+
+@WebListener
+public class InitListener implements ServletContextListener{
+        private static final Logger log = LoggerFactory.getLogger(InitListener.class);
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/web/controllers/listeners/InitListener.java]: you know with lombok you have @Sl4j...

**Review by SMishAlex — COMMENTED** [review]
don't close this merge, just resolve all issue and commit it here. I'll merge it into my brunch cause I don't wanna make review of all code every time, I'd prefer review only changes. and look after your naming

File: src/main/java/com/github/admissionCommittee/dao/DaoFactory.java (hunk 4 above)
Comment (kmavropulo) [reply src/main/java/com/github/admissionCommittee/dao/DaoFactory.java:13]: Like this:
...
private static ExamCertificateDaoInstance examCertificateDaoInstance;
public ExamCertificateDaoInstance getExamCertificateDaoInstance() 
...
private static volatile DaoFactory Instance ?

File: src/main/java/com/github/admissionCommittee/dao/GenericDao.java (hunk 5 above)
Comment (kmavropulo) [reply src/main/java/com/github/admissionCommittee/dao/GenericDao.java:64]: fixed:
public void delete(long id) {
        openSessionWithTransaction();
        T instance = session.get(type, id);
        if (instance != null) {
            session.delete(instance);
            session.flush();
            closeSession();
        }
    }

File: src/main/java/com/github/admissionCommittee/model/User.java [hunk 9]
```diff
@@ -63,6 +63,19 @@
...
+    @Column(name = "is_enlisted")
+    private boolean isEnlisted;
+
+    //TODO service check
+
+    public boolean isEnlisted() {
+        return isEnlisted;
+    }
+
+    public void setEnlisted(boolean enlisted) {
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/model/User.java]: if you call your field just `enlisted` lombok will generate the same code

File: src/main/java/com/github/admissionCommittee/dao/GenericDao.java (hunk 5 above)
Comment (SMishAlex) [reply src/main/java/com/github/admissionCommittee/dao/GenericDao.java:64]: close session out of `if` or you may have unclose session

File: src/main/java/com/github/admissionCommittee/service/UserService.java [hunk 10]
```diff
@@ -15,5 +28,56 @@ public User getByMail(String mail) {
         return ((UserDao) getDao()).getByMail(mail);
     }
 
+    public void updateIsEnlisted(Map<Faculty, List<User>> facultyAttendeeList) {
+        facultyAttendeeList.values().parallelStream()
+                .forEachOrdered(list -> list.parallelStream().forEachOrdered(user -> user
+                        .setEnlisted(true)));
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/service/UserService.java]: parallelStream gives benefits only on big set of data (more than 10k), if you care about speed then you should use forEach without order, if I were you I'd use something like `stream().flatMap(list->list.stream()).forEach(..)`. also forEach is terminate method with side-effect and you should avoid use it if you  work in functional-base style

File: src/main/java/com/github/admissionCommittee/dao/GenericDao.java [hunk 11]
```diff
@@ -16,15 +17,25 @@ public GenericDao(Class<T> type) {
...
+        if (instance.isNew()) {
+            session.save(instance);
+        } else {
+            session.update(instance);
+        }
         closeSessionWithTransaction();
     }
 
-    public void update(T instance) {
+    public void save(List<T> instance) {
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/dao/GenericDao.java]: usually param of this method has Iterable<T> type

File: src/main/java/com/github/admissionCommittee/service/SheetService.java [hunk 12]
```diff
@@ -1,22 +1,54 @@
...
+
+    private int calculateExamScoreSum(Sheet instance) {
+        final User user = instance.getUser();
+        final Faculty faculty = instance.getFaculty();
+        final ExamCertificate userExamCertificate = user.getExamCertificate();
+        final Map<Subject, Integer> examScoreMap = userExamCertificate.getSubjects();
+        final Set<Subject> facultySubjects = faculty.getSubjects();
+        return examScoreMap.keySet().stream()
+            .filter(facultySubjects::contains)
+            .map(examScoreMap::get)
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/service/SheetService.java]: i suppose it's better to use mapToInt and IntStream has method sum

**Review by SMishAlex — COMMENTED** [review]
please pay attention on parallelStream, as I've wrote before: it brings benefits on huge set of data, and don't use forEachOrdered if you don't care about order

File: src/main/java/com/github/admissionCommittee/dao/DaoFactory.java (hunk 4 above)
Comment (SMishAlex) [reply src/main/java/com/github/admissionCommittee/dao/DaoFactory.java:13]: I don't really get why you use lazy init of all this instances... but you get the Idea correctly

**Review by SMishAlex — COMMENTED** [review]
I haven't any suggestions

File: src/main/java/com/github/admissionCommittee/model/User.java [hunk 13]
```diff
@@ -25,13 +26,17 @@
 @NoArgsConstructor
 @Table(name = "user")
 @EqualsAndHashCode(callSuper = true, doNotUseGetters = true)
-@AttributeOverride(name = "id", column = @Column(name = "user_id", nullable = false))
+@AttributeOverride(name = "id", column = @Column(name = "user_id", nullable =
+        false))
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/model/User.java]: it's not so long to split the line

File: src/main/java/com/github/admissionCommittee/service/ServiceFactory.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+import com.github.admissionCommittee.model.AbstractEntity;
+import com.github.admissionCommittee.model.ExamCertificate;
+import com.github.admissionCommittee.model.Faculty;
+import com.github.admissionCommittee.model.SchoolCertificate;
+import com.github.admissionCommittee.model.Sheet;
+import com.github.admissionCommittee.model.Subject;
+import com.github.admissionCommittee.model.User;
+
+;
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/service/ServiceFactory.java]: wtf?

File: src/main/java/com/github/admissionCommittee/util/validate/UserValidatorUtil.java [hunk 15]
```diff
@@ -2,35 +2,77 @@
...
+
 
 public class UserValidatorUtil extends ValidatorUtil {
+    private static final Logger log = LoggerFactory.getLogger
+            (UserValidatorUtil.class);
+
     @Override
-    public void validateEntity(AbstractEntity entityToValidate) {
+    public void validate(AbstractEntity entityToValidate) {
         User toValidate = (User) entityToValidate;
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/util/validate/UserValidatorUtil.java]: class cast here is really bad solution... I suppouse it's better to make your `ValidatorUtil` paramitrise (like ` ValidatorUtil<T extends AbstractEntity>`

File: src/main/java/com/github/admissionCommittee/service/UserService.java [hunk 16]
```diff
@@ -22,24 +23,30 @@
...
+    @Override
+    public void save(User instance) {
+        validator.validate(instance);
+        super.save(instance);
+    }
+
+    @Override
+    public void save(List<User> instance) {
+        instance.forEach(validator::validate);
+        super.save(instance);
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/service/UserService.java]: if at list one user is not valid this method will not save any

File: src/main/java/com/github/admissionCommittee/util/init/FacultyInitializerUtil.java [hunk 17]
```diff
@@ -14,41 +18,65 @@
...
+            final int[] counter = {0};
             Files.lines(Paths.get(inputFile))
                     .map(line -> line.split("\u200B"))
                     .forEach(substring -> {
                         Faculty faculty = new Faculty(substring[2], Integer
                                 .parseInt(substring[3]), getRandomSubject(
                                 3));
-                        ValidatorUtil.getValidator(Faculty.class)
-                                .validateEntity(faculty);
+                        validator.validate(faculty);
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/util/init/FacultyInitializerUtil.java]: unchecked exception without processing

File: src/main/java/com/github/admissionCommittee/util/init/SchoolCertificateInitializerUtil.java [hunk 18]
```diff
@@ -1,15 +1,54 @@
...
+                .forEach(user -> {
+                    SchoolCertificate schoolCertificate = new
+                            SchoolCertificate(user, user.getBirthDate()
+                            .getYear() + 17,
+                            ScoresUtil.getRandomScores(subjectList, 3, 2));
+                    validator.validate(schoolCertificate);
+                    schoolCertificates.add(schoolCertificate);
+                    //assign school certificate to user
+                    user.setSchoolCertificate(schoolCertificate);
+                    counter[0]++;
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/util/init/SchoolCertificateInitializerUtil.java]: too many side effects in one lambda...

File: src/main/java/com/github/admissionCommittee/util/init/SheetInitializerUtil.java [hunk 19]
```diff
@@ -1,14 +1,63 @@
...
+                                    user.getExamCertificate()
+                                            .getSubjects()) / user
+                                    .getExamCertificate()
+                                    .getSubjects()
+                                    .size());
+                    validator.validate(sheet);
+                    sheets.add(sheet);
+                    //assign sheet to user
+                    user.setSheet(sheet);
+                    counter[0]++;
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/util/init/SheetInitializerUtil.java]: don't use functional programming if you don't know how to use it

File: src/main/java/com/github/admissionCommittee/util/validate/org/apache/commons/validator/routines/RegexValidator.java [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+        buffer.append("RegexValidator{");
+        for (int i = 0; i < patterns.length; i++) {
+            if (i > 0) {
+                buffer.append(",");
+            }
+            buffer.append(patterns[i].pattern());
+        }
+        buffer.append("}");
+        return buffer.toString();
+    }
```
Comment (SMishAlex) [inline src/main/java/com/github/admissionCommittee/util/validate/org/apache/commons/validator/routines/RegexValidator.java:239]: should I review this lib?

File: src/main/java/com/github/admissionCommittee/service/UserService.java (hunk 16 above)
Comment (kmavropulo) [reply src/main/java/com/github/admissionCommittee/service/UserService.java]: This method is for initialization and testing only, therefore, if validation fails, we just need to fix the validation algorithm or input data (if fault is in input). 
For production use, we can easily overload the method with filter.

Comment (SMishAlex) [reply src/main/java/com/github/admissionCommittee/service/UserService.java]: there is no documentation about that, so for me and other programmers its public method of CRUDrepository which we always use in production not only for initialization, also there is no throws segment and if thes method will thow sth that kill my site I will be very surprised

File: src/main/java/com/github/admissionCommittee/util/validate/UserValidatorUtil.java (hunk 15 above)
Comment (kmavropulo) [reply src/main/java/com/github/admissionCommittee/util/validate/UserValidatorUtil.java]: ValidatorUtil using for many classes's validation.

File: src/main/java/com/github/admissionCommittee/util/init/FacultyInitializerUtil.java (hunk 17 above)
Comment (kmavropulo) [reply src/main/java/com/github/admissionCommittee/util/init/FacultyInitializerUtil.java]: Validation implemented, during init logging to file; all services save method's returns logs of errors; validation of process of user's profile creation based on it (checking the password, name, etc, if failes, user see it in the message).

File: src/main/java/com/github/admissionCommittee/service/UserService.java (hunk 16 above)
Comment (kmavropulo) [reply src/main/java/com/github/admissionCommittee/service/UserService.java]: Documentation added.

File: src/main/java/com/github/admissionCommittee/util/init/SheetInitializerUtil.java (hunk 19 above)
Comment (kmavropulo) [reply src/main/java/com/github/admissionCommittee/util/init/SheetInitializerUtil.java]: Fixed.

File: src/main/java/com/github/admissionCommittee/util/init/SchoolCertificateInitializerUtil.java (hunk 18 above)
Comment (kmavropulo) [reply src/main/java/com/github/admissionCommittee/util/init/SchoolCertificateInitializerUtil.java]: Fixed.

File: src/main/java/com/github/admissionCommittee/service/ServiceFactory.java (hunk 14 above)
Comment (kmavropulo) [reply src/main/java/com/github/admissionCommittee/service/ServiceFactory.java]: Fixed.

**Pull request merged**